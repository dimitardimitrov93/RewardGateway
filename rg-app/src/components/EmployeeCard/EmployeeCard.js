import * as React from 'react';

import Card from '@mui/material/Card';
import CardContent from '@mui/material/CardContent';
import CardHeader from '@mui/material/CardHeader';
import Avatar from '@mui/material/Avatar';
import IconButton from '@mui/material/IconButton';
import { styled } from '@mui/material/styles';
import TextField from '@mui/material/TextField';
import validator from 'validator';
import ColorRadioButtons from '../ColorRadioButtons/ColorRadioButtons';
import Collapse from '@mui/material/Collapse';
import ExpandMoreIcon from '@mui/icons-material/ExpandMore';

import Typography from '@mui/material/Typography';
import { Button, CardActionArea, CardActions } from '@mui/material';

const ExpandMore = styled((props) => {
    const { expand, ...other } = props;
    return <IconButton {...other} />;
})(({ theme, expand }) => ({
    transform: !expand ? 'rotate(0deg)' : 'rotate(180deg)',
    marginLeft: 'auto',
    transition: theme.transitions.create('transform', {
        duration: theme.transitions.duration.shortest,
    }),
}));

export default function EmployeeCard({ employee, handleUpdateEmployee }) {
    const [label, setLabel] = React.useState('');
    const [currentLabelField, setCurrentLabelField] = React.useState(null);
    const [backgroundColor, setBackgroundColor] = React.useState('');

    const [expanded, setExpanded] = React.useState(false);

    const sanitizedEmployeeInfo = {};

    for (const key in employee) {
        if (key !== 'avatar') {
            sanitizedEmployeeInfo[key] = validator.escape(sanitizeHTML(employee[key]));
        } else {
            sanitizedEmployeeInfo[key] = employee[key];
        }
    }

    const handleExpandClick = () => {
        setExpanded(!expanded);
    };

    function sanitizeHTML(htmlText) {
        const scriptRegExp = /<.*script.*>.*<.*\/.*script.*>/gi;
        const htmlCharsRegex = /&amp;|nbsp;|amp;|&lt;|&gt;|gt;|lt;/gi;
        const tempHtmlText = htmlText.replace(scriptRegExp, '');
        const sanitizedHtmlText = tempHtmlText.replace(htmlCharsRegex, '');
        return sanitizedHtmlText;
    }

    function handleChange(e) {
        e.preventDefault();
        setLabel(e.target.value);
        setCurrentLabelField(e.target);
    }

    function onSubmitLabel() {
        if (!label) return;

        const sanitizedLabel = validator.escape(label);
        setLabel(sanitizedLabel);
        // sanitizedEmployeeInfo[backgroundColor] = backgroundColor;

        handleUpdateEmployee({ ...sanitizedEmployeeInfo, label });

        setLabel('');
        currentLabelField.value = '';
        currentLabelField.blur();
    }

    function onChangeColorSubmit() {
        handleUpdateEmployee({ ...sanitizedEmployeeInfo, backgroundColor });
        // handleExpandClick()
    }

    return (
        <Card sx={{ maxWidth: 450, backgroundColor: sanitizedEmployeeInfo.backgroundColor }}>
            {sanitizedEmployeeInfo.label &&
                <Typography gutterBottom variant="h5" component="div" sx={{ paddingTop: '0.3em' }}>
                    Label: {sanitizedEmployeeInfo.label}
                </Typography>
            }
            <CardActionArea>
                <a href={sanitizedEmployeeInfo.avatar} target="_blank">
                    <CardHeader
                        data-testid="cardHeader-1"
                        avatar={
                            <Avatar
                                aria-label="avatar"
                                src={sanitizedEmployeeInfo.avatar}
                                alt="person's avatar image"
                            >
                                {sanitizedEmployeeInfo.name[0].toUpperCase()}
                            </Avatar>
                        }
                        title={sanitizedEmployeeInfo.name}
                        subheader={sanitizedEmployeeInfo.title}
                        sx={{ '.MuiCardHeader-title': { fontWeight: "bold" } }}
                    />
                </a>
            </CardActionArea>
            <CardContent>
                <Typography data-testid="typography-1" gutterBottom variant="h5" component="div">
                    {sanitizedEmployeeInfo.company}
                </Typography>
                <Typography data-testid="typography-2" color="text.secondary" sx={{ padding: '1em', textAlign: 'justify' }}>
                    {sanitizedEmployeeInfo.bio}
                </Typography>
                <CardActions>
                    <TextField
                        id="standard-basic"
                        label="Label"
                        variant="standard"
                        onChange={handleChange}
                    />
                    <Button sx={{ margin: '1.2em 0 0 1em' }} variant="contained" onClick={onSubmitLabel} size="small" color="primary">
                        Add
                        </Button>

                </CardActions>
            </CardContent>
            <ExpandMore
                expand={expanded}
                onClick={handleExpandClick}
                aria-expanded={expanded}
                aria-label="show more"
            >
                <ExpandMoreIcon />
            </ExpandMore>
            <Collapse in={expanded} timeout="auto" unmountOnExit>
                <CardContent>
                    <div>
                        <ColorRadioButtons setBackgroundColor={setBackgroundColor} />
                        <Button onClick={onChangeColorSubmit} size="small" color="primary">
                            Change Color
                        </Button>
                    </div>
                </CardContent>
            </Collapse>

        </Card>
    );
}