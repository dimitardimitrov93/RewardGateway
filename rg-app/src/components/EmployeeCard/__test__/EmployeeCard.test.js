import React from 'react';
import ReactDOM from 'react-dom';
import EmployeeCard from '../EmployeeCard';
import { render, screen, cleanup } from '@testing-library/react';
import renderer from 'react-test-renderer';

const employee = {
    "uuid": "55a3b557-971f-3e84-9870-5e1bdc5a12d3",
    "company": "Ferry-Dach",
    "bio": "Id qui repellendus eveniet eos odit ea autem. Ipsum veniam dicta omnis. Quae laudantium aperiam voluptas impedit veniam totam repudiandae pariatur. Et non eum aut.",
    "name": "Francesco Green",
    "title": "Production Control Manager",
    "avatar": "https://lorempixel.com/64/64/people/?75721"
}

it('renders EmployeeCard without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<EmployeeCard employee={employee} />, div);
    ReactDOM.unmountComponentAtNode(div);
});

it('renders Typography component with employee company info correctly', () => {
    render(<EmployeeCard employee={employee} />);
    const TypographyElement = screen.getByTestId('typography-1');
    expect(TypographyElement).toBeInTheDocument();
});

it('renders Typography component with employee company info correctly', () => {
    render(<EmployeeCard employee={employee} />);
    const TypographyElement = screen.getByTestId('typography-1');
    expect(TypographyElement).toBeInTheDocument();
});

it('renders Typography component with employee bio correctly', () => {
    render(<EmployeeCard employee={employee} />);
    const TypographyElement = screen.getByTestId('typography-2');
    expect(TypographyElement).toBeInTheDocument();
    expect(TypographyElement).toHaveTextContent(employee.bio);
});

it('renders CardHeader component with employee name and title correctly', () => {
    render(<EmployeeCard employee={employee} />);
    const cardHeaderElement = screen.getByTestId('cardHeader-1');
    expect(cardHeaderElement).toBeInTheDocument();
    expect(cardHeaderElement).toHaveTextContent(employee.name);
    expect(cardHeaderElement).toHaveTextContent(employee.title);
});

test('matches snapshot', () => {
    const tree = renderer.create(<EmployeeCard employee={employee} />).toJSON();
    expect(tree).toMatchSnapshot();
});

afterEach(() => {
    cleanup();
});