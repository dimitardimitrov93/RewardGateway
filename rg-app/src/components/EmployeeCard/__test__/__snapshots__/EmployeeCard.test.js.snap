// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1d9ctjl-MuiPaper-root-MuiCard-root"
>
  <button
    className="MuiButtonBase-root MuiCardActionArea-root css-b6lsxj-MuiButtonBase-root-MuiCardActionArea-root"
    disabled={false}
    onBlur={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <a
      href="https://lorempixel.com/64/64/people/?75721"
      target="_blank"
    >
      <div
        className="MuiCardHeader-root css-9vkofw-MuiCardHeader-root"
        data-testid="cardHeader-1"
      >
        <div
          className="MuiCardHeader-avatar css-1ssile9-MuiCardHeader-avatar"
        >
          <div
            aria-label="avatar"
            className="MuiAvatar-root MuiAvatar-circular css-1wlk0hk-MuiAvatar-root"
          >
            <img
              alt="person's avatar image"
              className="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
              src="https://lorempixel.com/64/64/people/?75721"
            />
          </div>
        </div>
        <div
          className="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
        >
          <span
            className="MuiTypography-root MuiTypography-body2 MuiCardHeader-title css-et1ao3-MuiTypography-root"
          >
            Francesco Green
          </span>
          <span
            className="MuiTypography-root MuiTypography-body2 MuiCardHeader-subheader css-83ijpv-MuiTypography-root"
          >
            Production Control Manager
          </span>
        </div>
      </div>
    </a>
    <span
      className="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
    />
  </button>
  <div
    className="MuiCardContent-root css-46bh2p-MuiCardContent-root"
  >
    <div
      className="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
      data-testid="typography-1"
    >
      Ferry-Dach
    </div>
    <p
      className="MuiTypography-root MuiTypography-body1 css-lrzl8a-MuiTypography-root"
      data-testid="typography-2"
    >
      Id qui repellendus eveniet eos odit ea autem. Ipsum veniam dicta omnis. Quae laudantium aperiam voluptas impedit veniam totam repudiandae pariatur. Et non eum aut.
    </p>
    <div
      className="MuiCardActions-root MuiCardActions-spacing css-1t6e9jv-MuiCardActions-root"
    >
      <div
        className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      >
        <label
          className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard MuiFormLabel-root MuiFormLabel-colorPrimary css-aqpgxn-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink={false}
          htmlFor="standard-basic"
          id="standard-basic-label"
        >
          Label
        </label>
        <div
          className="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-1480iag-MuiInputBase-root-MuiInput-root"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInput-input MuiInputBase-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
            disabled={false}
            id="standard-basic"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="text"
          />
        </div>
      </div>
      <button
        className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButtonBase-root css-1c38yg3-MuiButtonBase-root-MuiButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        Add
      </button>
    </div>
  </div>
  <button
    aria-expanded={false}
    aria-label="show more"
    className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-3ogxn6-MuiButtonBase-root-MuiIconButton-root"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="ExpandMoreIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
      />
    </svg>
  </button>
</div>
`;
